op-monitorism/transaction_monitor/monitor.go

```diff
@@ -0,0 +1,411 @@
+					verifier := &DisputeGameVerifier{
+						factory:    factoryAddr,
+						factoryABI: factoryABI,
+						gameABI:    gameABI,
+						cache:      make(map[common.Address]bool),
+					}
+					mon.gameVerifiers[factoryAddr] = verifier
+				}
+			}
+		}
+	}
+
+	return mon, nil
+}
+
+func (v *DisputeGameVerifier) verifyGame(ctx context.Context, client *ethclient.Client, gameAddr common.Address) (bool, error) {
+	v.mu.RLock()
+	if cached, exists := v.cache[gameAddr]; exists {
+		v.mu.RUnlock()
+		return cached, nil
+	}
+	v.mu.RUnlock()
+
+	// Create contract bindings
+	game := bind.NewBoundContract(gameAddr, v.gameABI, client, client, client)
+	factory := bind.NewBoundContract(v.factory, v.factoryABI, client, client, client)
+
+	// Get game parameters
+	var gameTypeResult []interface{}
+	if err := game.Call(&bind.CallOpts{Context: ctx}, &gameTypeResult, "gameType"); err != nil {
+		return false, fmt.Errorf("failed to get game type: %w", err)
+	}
+	if len(gameTypeResult) != 1 {
+		return false, fmt.Errorf("unexpected number of return values for gameType")
+	}
+	gameType, ok := gameTypeResult[0].(uint32)
+	if !ok {
+		return false, fmt.Errorf("invalid game type returned")
+	}
+
+	var rootClaimResult []interface{}
+	if err := game.Call(&bind.CallOpts{Context: ctx}, &rootClaimResult, "rootClaim"); err != nil {
+		return false, fmt.Errorf("failed to get root claim: %w", err)
+	}
+	if len(rootClaimResult) != 1 {
+		return false, fmt.Errorf("unexpected number of return values for rootClaim")
+	}
+	rootClaim, ok := rootClaimResult[0].([32]byte)
+	if !ok {
+		return false, fmt.Errorf("invalid root claim returned")
+	}
+
+	var extraDataResult []interface{}
+	if err := game.Call(&bind.CallOpts{Context: ctx}, &extraDataResult, "extraData"); err != nil {
+		return false, fmt.Errorf("failed to get extra data: %w", err)
+	}
+	if len(extraDataResult) != 1 {
+		return false, fmt.Errorf("unexpected number of return values for extraData")
+	}
+	extraData, ok := extraDataResult[0].([]byte)
+	if !ok {
+		return false, fmt.Errorf("invalid extra data returned")
+	}
+
+	// Verify with factory
+	var factoryResult []interface{}
+	if err := factory.Call(&bind.CallOpts{Context: ctx}, &factoryResult, "games", gameType, rootClaim, extraData); err != nil {
+		return false, fmt.Errorf("failed to verify game with factory: %w", err)
+	}
+	if len(factoryResult) != 2 {
+		return false, fmt.Errorf("unexpected number of return values from factory")
+	}
+
+	proxy, ok := factoryResult[0].(common.Address)
+	if !ok {
+		return false, fmt.Errorf("invalid proxy address returned")
+	}
+
+	timestamp, ok := factoryResult[1].(uint64)
+	if !ok {
+		return false, fmt.Errorf("invalid timestamp returned")
+	}
+
+	isValid := proxy == gameAddr && timestamp > 0
+
+	// Cache the result
+	v.mu.Lock()
+	v.cache[gameAddr] = isValid
+	v.mu.Unlock()
+
+	return isValid, nil
+}
+
+func (m *Monitor) Run(ctx context.Context) {
+	// Initialize starting block
+	currentBlock := m.startBlock
```

Comment:
Should track the current block number as a metric so that we can know if the monitor is stalled for some reason